tests/src/python/test_qgsserver_accesscontrol.py

```diff
@@ -0,0 +1,1082 @@
+        self.assertTrue(
+            str(response).find("<qgs:pk>") != -1,
+            "No result in GetFeature\n%s" % response)
+        self.assertTrue(
+            str(response).find("<qgs:pk>1</qgs:pk>") != -1,
+            "No good result in GetFeature\n%s" % response)
+
+    def test_wfs_getfeature_subsetstring2(self):
+        data = """<?xml version="1.0" encoding="UTF-8"?>
+            <wfs:GetFeature {xml_ns}>
+            <wfs:Query typeName="Hello_SubsetString" srsName="EPSG:3857" xmlns:feature="http://www.qgis.org/gml">
+            <ogc:Filter xmlns:ogc="http://www.opengis.net/ogc"><ogc:PropertyIsEqualTo>
+            <ogc:PropertyName>pkuid</ogc:PropertyName>
+            <ogc:Literal>2</ogc:Literal>
+            </ogc:PropertyIsEqualTo></ogc:Filter></wfs:Query></wfs:GetFeature>""".format(xml_ns=XML_NS)
+
+        response, headers = self._post_fullaccess(data)
+        self.assertTrue(
+            str(response).find("<qgs:pk>") != -1,
+            "No result in GetFeature\n%s" % response)
+        self.assertTrue(
+            str(response).find("<qgs:pk>2</qgs:pk>") != -1,
+            "No good result in GetFeature\n%s" % response)
+
+        response, headers = self._post_restricted(data)
+        self.assertTrue(
+            str(response).find("<qgs:pk>") == -1,
+            "Unexpected result in GetFeature\n%s" % response)
+
+    def _handle_request(self, restricted, *args):
+        accesscontrol.restricted = restricted
+        result = self._result(server.handleRequest(*args))
+        return result
+
+    def _result(self, data):
+        pos = str(data).find("\n\n")
+        headers = {}
+        for line in data[:pos].split("\n"):
+            header = line.split(":")
+            self.assertEquals(len(header), 2, line)
+            headers[str(header[0])] = str(header[1]).strip()
+
+        return data[pos+2:], headers
+
+    def _get_fullaccess(self, query_string):
+        os.environ["REQUEST_METHOD"] = "GET"
+        result = self._handle_request(False, query_string)
+        del os.environ["REQUEST_METHOD"]
+        return result
+
+    def _get_restricted(self, query_string):
+        os.environ["REQUEST_METHOD"] = "GET"
+        result = self._handle_request(True, query_string)
+        del os.environ["REQUEST_METHOD"]
+        return result
+
+    def _post_fullaccess(self, data, query_string=None):
+        os.environ["REQUEST_METHOD"] = "POST"
+        os.environ["REQUEST_BODY"] = data
+        result = self._handle_request(False, query_string)
+        del os.environ["REQUEST_METHOD"]
+        del os.environ["REQUEST_BODY"]
+        return result
+
+    def _post_restricted(self, data, query_string=None):
+        os.environ["REQUEST_METHOD"] = "POST"
+        os.environ["REQUEST_BODY"] = data
+        result = self._handle_request(True, query_string)
+        del os.environ["REQUEST_METHOD"]
+        del os.environ["REQUEST_BODY"]
+        return result
+
+    def _img_diff(self, image_1, image_2):
```

Comment:
Please use the existing QgsRenderChecker class for testing rendered images - it allows us to put in place various required fixes like platform specific reference images and test masks.